File: server/app/stats.py
"""Pure functions that derive statistics from stored games.

No storage or framework concerns here — just the hockey math. This keeps the
processing logic easy to test and reason about.
"""

from __future__ import annotations

from .models import CareerStats, EventType, Game, GameStats


def _count(game: Game, event_type: EventType) -> int:
    return sum(1 for e in game.events if e.type == event_type)


def compute_game_stats(game: Game) -> GameStats:
    goals = _count(game, EventType.goal)
    assists = _count(game, EventType.assist)
    shots = _count(game, EventType.shot)
    faceoff_wins = _count(game, EventType.faceoff_win)
    faceoff_losses = _count(game, EventType.faceoff_loss)
    faceoffs = faceoff_wins + faceoff_losses

    total_ice = sum(s.durationSeconds for s in game.shifts)
    shift_count = len(game.shifts)

    return GameStats(
        gameId=game.id,
        goals=goals,
        assists=assists,
        points=goals + assists,
        shots=shots,
        shootingPct=(goals / shots) if shots else 0.0,
        hits=_count(game, EventType.hit),
        blocks=_count(game, EventType.block),
        penalties=_count(game, EventType.penalty),
        faceoffPct=(faceoff_wins / faceoffs) if faceoffs else 0.0,
        totalIceTimeSeconds=total_ice,
        shiftCount=shift_count,
        avgShiftSeconds=(total_ice / shift_count) if shift_count else 0.0,
        biometrics=game.biometrics,
    )


def compute_career_stats(games: list[Game]) -> CareerStats:
    per_game = [compute_game_stats(g) for g in games]
    n = len(per_game)

    total_goals = sum(s.goals for s in per_game)
    total_assists = sum(s.assists for s in per_game)
    total_shots = sum(s.shots for s in per_game)
    total_points = total_goals + total_assists
    total_ice = sum(s.totalIceTimeSeconds for s in per_game)

    return CareerStats(
        gamesPlayed=n,
        totalGoals=total_goals,
        totalAssists=total_assists,
        totalPoints=total_points,
        totalShots=total_shots,
        shootingPct=(total_goals / total_shots) if total_shots else 0.0,
        pointsPerGame=(total_points / n) if n else 0.0,
        avgIceTimeSeconds=(total_ice / n) if n else 0.0,
    )

File: shared/Sources/HockeyContract/Stats.swift
import Foundation

/// Mirrors `#/components/schemas/GameStats`.
/// Derived numbers for a single game — computed by the server; clients display.
public struct GameStats: Codable, Equatable {
    public let gameId: String
    public let goals: Int
    public let assists: Int
    public let points: Int
    public let shots: Int
    /// goals / shots, 0..1.
    public let shootingPct: Double
    public let hits: Int
    public let blocks: Int
    public let penalties: Int
    /// wins / (wins + losses), 0..1.
    public let faceoffPct: Double
    public let totalIceTimeSeconds: Double
    public let shiftCount: Int
    /// Optional in the contract.
    public let avgShiftSeconds: Double?
    public let biometrics: BiometricSummary?

    public init(
        gameId: String,
        goals: Int,
        assists: Int,
        points: Int,
        shots: Int,
        shootingPct: Double,
        hits: Int,
        blocks: Int,
        penalties: Int,
        faceoffPct: Double,
        totalIceTimeSeconds: Double,
        shiftCount: Int,
        avgShiftSeconds: Double? = nil,
        biometrics: BiometricSummary? = nil
    ) {
        self.gameId = gameId
        self.goals = goals
        self.assists = assists
        self.points = points
        self.shots = shots
        self.shootingPct = shootingPct
        self.hits = hits
        self.blocks = blocks
        self.penalties = penalties
        self.faceoffPct = faceoffPct
        self.totalIceTimeSeconds = totalIceTimeSeconds
        self.shiftCount = shiftCount
        self.avgShiftSeconds = avgShiftSeconds
        self.biometrics = biometrics
    }
}

/// Mirrors `#/components/schemas/CareerStats`.
/// Totals and per-game averages across every game — computed by the server.
public struct CareerStats: Codable, Equatable {
    public let gamesPlayed: Int
    public let totalGoals: Int
    public let totalAssists: Int
    public let totalPoints: Int
    /// Optional in the contract.
    public let totalShots: Int?
    /// Optional in the contract.
    public let shootingPct: Double?
    public let pointsPerGame: Double
    /// Optional in the contract.
    public let avgIceTimeSeconds: Double?

    public init(
        gamesPlayed: Int,
        totalGoals: Int,
        totalAssists: Int,
        totalPoints: Int,
        totalShots: Int? = nil,
        shootingPct: Double? = nil,
        pointsPerGame: Double,
        avgIceTimeSeconds: Double? = nil
    ) {
        self.gamesPlayed = gamesPlayed
        self.totalGoals = totalGoals
        self.totalAssists = totalAssists
        self.totalPoints = totalPoints
        self.totalShots = totalShots
        self.shootingPct = shootingPct
        self.pointsPerGame = pointsPerGame
        self.avgIceTimeSeconds = avgIceTimeSeconds
    }
}
